rasterio/_warp.pyx

```diff
@@ -455,8 +456,19 @@ def _reproject(
     # Set up GDALCreateGenImgProjTransformer2 keyword arguments.
     cdef char **imgProjOptions = NULL
     imgProjOptions = CSLSetNameValue(imgProjOptions, "GCPS_OK", "TRUE")
+
     if rpcs:
         imgProjOptions = CSLSetNameValue(imgProjOptions, "SRC_METHOD", "RPC")
+    elif src_geoloc_array is not None:
+        arr = np.stack((src_geoloc_array[0], src_geoloc_array[1]))
```

Comment:
As mentioned above, a geolocation array can also have a Z band, but I suspect it is ignored by `GDALWapr()`.